rtio: add helper function `rtio_read_transaction()`

avisconti opened the pull request:
Add a helper function that constructs a rtio SQE chain with the purpose to perform a bus read operation on a list of registers.

Tested with lsm6dsv16x sensor:

I2C bus: using nucleo_h503rb board + x_nucleo_iks4a1 shield
SPI bus: using sensortile_box_pro

```
west build -b nucleo_h503rb samples/sensor/stream_fifo/
west build -b nucleo_h503rb  samples/sensor/stream_drdy/
west build -b sensortile_box_pro samples/sensor/stream_fifo/
west build -b sensortile_box_pro samples/sensor/stream_drdy/
```
(note that to test with I2C I had to use #91664)

Helper usage:

```
       struct rtio_regs regs;
       struct rtio_reg_list regs_list[] = {{regs_addr1, mem_addr_1, mem_len_1},
                                           {regs_addr2, mem_addr_2, mem_len_2},
                                           ...
                                          };
       regs.rtio_regs_list = regs_list;
       regs.rtio_regs_num = ARRAY_SIZE(regs_list);
    
       rtio_read_regs_async(rtio,
                            iodev,
                            RTIO_BUS_SPI,
                            &regs,
                            sqe,
                            dev,
                            op_cb);
```

Points to be discussed:

1. This helper is defined as static inline. Since it may be called in more places by several drivers, this would lead to a code size increasing. Is there any other possibility?
2. The helper currently is declared as void. Should we consider returning an error code?
3. I'm handling the bus part internally, because, in my opinion, this part seems to pertain to RTIO-enabled bus. Is this matching what we have said in the sensor WG? I may have missed some comments...
4. This helper would fail if the rtio resources are not properly dimensioned in the DT. This is true always, even if the helper is not there. But maybe we should properly use ASSERT to help implementors to debug...

avisconti committed:
rtio: add helper function `rtio_read_transaction()`

Add a helper function that constructs a rtio SQE chain with the purpose
to perform a bus read operation on a list of registers.

Usage:

  struct rtio_regs regs;
  struct rtio_reg_list regs_list[] = {{regs_addr1, mem_addr_1, mem_len_1},
                                      {regs_addr2, mem_addr_2, mem_len_2},
                                      ...
                                     };
  regs.rtio_regs_list = regs_list;
  regs.rtio_regs_num = ARRAY_SIZE(regs_list);

  rtio_read_regs_async(rtio,
                       iodev,
                       RTIO_BUS_SPI,
                       &regs,
                       sqe,
                       dev,
                       op_cb);

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
Is this PR matching your expectation?

teburd reviewed (commented):
I think this makes sense, especially as there's already a need in some st drivers. Having it here means we can always update it with a grep and refactor if needed.

+1 from me once the function params are sorted out

avisconti commented:
I rebased just to have https://github.com/zephyrproject-rtos/zephyr/pull/91664 included.
Reworked the arguments order as per @teburd suggestions.

Still, I have some doubts on this point:
"This helper is defined as static inline. Since it may be called in more places by several drivers, this would lead to a code size increasing. Is there any other possibility?"
Any consideration?

teburd commented on include/zephyr/rtio/rtio.h:
```diff
@@ -1674,6 +1713,91 @@ static inline int z_impl_rtio_submit(struct rtio *r, uint32_t wait_count)
+ * @param r RTIO context
+ * @param rtio_callback_t callback routine at the end of op
+ */
+static inline void rtio_read_transaction(const struct device *dev,
```
I missed this clearly! thanks for pointing it out

teburd commented:
> I rebased just to have #91664 included. Reworked the arguments order as per @teburd suggestions.
> 
> Still, I have some doubts on this point: "This helper is defined as static inline. Since it may be called in more places by several drivers, this would lead to a code size increasing. Is there any other possibility?" Any consideration?

Try it and see, what I've seen gcc do now is intelligently generate symbols for these functions when they don't meet the inline criteria, and otherwise inline them. So, I believe (and please tell me I'm wrong if so!) that the compiler inline heuristics are deciding what to inline and when rather than explicitly forcing a function call or always inlining (ALWAYS_INLINE isn't used here).

ubieda reviewed (commented):
Just dropping a drive-by comment (I can take a look once I'm back at the office next week, unless it's merged before):

I think we'll want to come back later and refactor this when we do introduce the regmap APIs.

Otherwise, I agree this helper is used in many places and it's useful.

ubieda commented on include/zephyr/rtio/rtio.h:
```diff
@@ -1674,6 +1713,91 @@ static inline int z_impl_rtio_submit(struct rtio *r, uint32_t wait_count)
+		}
+
+		if (rtio_is_spi(bus_type)) {
+			regs[i] |= 0x80; /* mark the SPI read transaction */
```
I feel like this should be configurable: there are some drivers that have this inverted (as in: 1 when writing, 0 when reading).

Here's one in-tree example: https://github.com/zephyrproject-rtos/zephyr/blob/4ec435165753bf353b35ae700c3ed5a1421eb46b/drivers/sensor/pixart/pat9136/pat9136_reg.h#L11-L12

avisconti replied:
I see. Yes, I remember now that you mentioned it during the WG call, but at the end I forgot it. So, the best thing to do is that the driver would raise the proper bit in ```regs_list[].reg_addr``` (regs_list is the second argument)

dkalowsk commented:
@avisconti please address the issues raised by @kartben and @ubieda so we can merge this for 4.2

avisconti commented:
> @avisconti please address the issues raised by @kartben and @ubieda so we can merge this for 4.2

Sorry, yesterday I was in vacation. I'll do it right now

avisconti commented on include/zephyr/rtio/rtio.h:
```diff
@@ -459,6 +459,45 @@ struct rtio {
+ * @brief A structure to describe a list of not-consecutive memory chunks
+ * for RTIO operations.
+ */
+struct rtio_reg_buf {
```
Do you mean only this one, or ALL the inline additions as well?

avisconti replied:
OK, I'll move ALL stuff in regmap.h

kartben reviewed (changes requested):
Need to move release notes changes to 4.3

kartben commented on include/zephyr/rtio/regmap.h:
```diff
@@ -0,0 +1,150 @@
+ * for RTIO operations.
+ */
+struct rtio_reg_buf {
+	/* Valid pointer to a data buffer */
+	uint8_t *bufp;
+
+	/* Length of the buffer in bytes */
+	size_t len;
```
If it's meant to be API then it needs to use Doxygen style comments. If not, make them "INTERNAL_HIDDEN"

kartben commented on include/zephyr/rtio/regmap.h:
```diff
@@ -0,0 +1,150 @@
+} rtio_bus_type;
+
+/**
+ * @brief chceck if bus is SPI
```
Doxygen docs need to be completed (`@param` and `@return` missing) - apply everywhere

kartben commented on include/zephyr/rtio/regmap.h:
```diff
@@ -0,0 +1,150 @@
+ *
+ * Usage:
+ *
+ * uint8_t regs_list[]  = { reg_addr1, reg_addr2. ... }
+ * struct rtio_reg_buf rbuf[] = {{mem_addr_1, mem_len_1},
+ *                               {mem_addr_2, mem_len_2},
+ *                               ...
+ *                              };
+ *
+ * rtio_read_regs_async(dev,
+ *                      regs_list,
+ *                      ARRAY_SIZE(regs_list),
+ *                      BUS_SPI,
+ *                      rbuf,
+ *                      sqe,
+ *                      iodev,
+ *                      rtio,
+ *                      op_cb);
```
Need to use `@code` block

avisconti replied:
ack

avisconti commented on include/zephyr/rtio/regmap.h:
```diff
@@ -0,0 +1,160 @@
+} rtio_bus_type;
+
+/**
+ * @brief chceck if bus is SPI
```
thanks for catching it

avisconti commented on include/zephyr/rtio/regmap.h:
```diff
@@ -0,0 +1,160 @@
+/*
+ * @brief Create a chain of SQEs representing a bus transaction to read a reg.
+ *
+ * The RTIO-enabled bus driver is isnstrumented to perform bus read ops
```
oooops, thx!

kartben reviewed (dismissed):
thanks!

teburd reviewed (dismissed):
Refresh my +1

Though the sonar rules should be met, this one in particular https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=91775&id=zephyrproject-rtos_zephyr&open=AZe4IbILJ9Mi5NPXwuJI is being discussed here #93624 as the sonar (misra) rule and checkpatch disagree it seems.

I'm inclined to believe the misra ruleset should supercede checkpatch but this needs to be I guess worked out.

avisconti commented:
> Instead consider the following which more helpfully and explicitly shows to a reader what is meant to happen when the bus is not i2c or i3c is in fact, nothing.
> 
> ```c
> switch(bus_type) {
> case RTIO_BUS_I2C:
>     /* i2c logic */
> case RTIO_BUS_I3C:
>     /* i3c logic */
> case RTIO_BUS_SPI:
> default:
>     break;
> }
> ```

I see your point. Using a switch with all cases implemented is in fact clearer to the reader what should happen. And maybe it would be easier to extend to a new bus in case. I'm reconsidering to re-implement in this way.

avisconti commented:
Thanks for all of your precious suggestions and comments

teburd reviewed (approved):
Sonar issue noting i is declared out of the loop, but unused outside of the loop seems legit at first glance but I might've missed it. It's not a significant concern of mine. Re-approving as many times as needed.

avisconti commented:
> Sonar issue noting i is declared out of the loop, but unused outside of the loop seems legit at first glance but I might've missed it. It's not a significant concern of mine. Re-approving as many times as needed.

I also fixed that issue. I think that it just needs to update the result.
